### PR TITLE
bpo-35972: _xxsubinterpreters: Fix potential integer truncation in channel_send()

### PR DESCRIPTION
**Pull request by izbyshev**
https://bugs.python.org/issue35972

### CODE REVIEW

**Comment by vstinner** [comment]
Since it's a private API and _xxsubinterpreters is a private module, I don't think that a NEWS entry is needed.

@ericsnowcurrently: Would you mind to review this fix?

**Review by vstinner — APPROVED** [review]
LGTM.

File: Lib/test/test__xxsubinterpreters.py [hunk 1]
```diff
@@ -393,7 +393,19 @@ def test_bytes(self):
                             for i in range(-1, 258))
 
     def test_int(self):
-        self._assert_values(range(-1, 258))
+        self._assert_values(itertools.chain(range(-1, 258),
+                                            [sys.maxsize, -sys.maxsize - 1]))
+
+    def test_non_shareable_int(self):
```
Comment (ericsnowcurrently) [inline Lib/test/test__xxsubinterpreters.py:399]: What is the result of this test on 32-bit platforms without the `pystate.c` changes?

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
(marking so it doesn't get merged yet)

File: Lib/test/test__xxsubinterpreters.py (hunk 1 above)
Comment (izbyshev) [reply Lib/test/test__xxsubinterpreters.py:399]:
```
======================================================================
FAIL: test_non_shareable_int (test.test__xxsubinterpreters.ShareableTypeTests) [2147483648]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/scratch2/izbyshev/cpython/Lib/test/test__xxsubinterpreters.py", line 408, in test_non_shareable_int
    interpreters.channel_send(self.cid, i)
AssertionError: OverflowError not raised

======================================================================
FAIL: test_non_shareable_int (test.test__xxsubinterpreters.ShareableTypeTests) [-2147483649]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/scratch2/izbyshev/cpython/Lib/test/test__xxsubinterpreters.py", line 408, in test_non_shareable_int
    interpreters.channel_send(self.cid, i)
AssertionError: OverflowError not raised

----------------------------------------------------------------------
```

And if I change `test_int` to use `sys.maxsize + 1` and `-sys.maxsize - 2`:
```
======================================================================
FAIL: test_int (test.test__xxsubinterpreters.ShareableTypeTests) [2147483648]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/scratch2/izbyshev/cpython/Lib/test/test__xxsubinterpreters.py", line 370, in _assert_values
    self.assertEqual(got, obj)
AssertionError: -2147483648 != 2147483648

======================================================================
FAIL: test_int (test.test__xxsubinterpreters.ShareableTypeTests) [-2147483649]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/scratch2/izbyshev/cpython/Lib/test/test__xxsubinterpreters.py", line 370, in _assert_values
    self.assertEqual(got, obj)
AssertionError: 2147483647 != -2147483649
```

Comment (ericsnowcurrently) [reply Lib/test/test__xxsubinterpreters.py:399]: Thanks for checking that.

**Comment by ericsnowcurrently** [comment]
As noted on the tracker issue, LGTM.  I have added a comment to `_long_shared()` about the consequence for 32-bit architectures.  I'll merge this in a little while.

**Review by ericsnowcurrently — APPROVED** [review]

**Comment by vstinner** [comment]
```
    /* Note that this means the size of shareable ints is bounded by
     * sys.maxsize.  Hence on 32-bit architectures that is half the
     * size of maximum shareable ints on 64-bit.
     */
```

If you really need 64-bit on 32-bit system, you can use int64_t. But it means that you should change _PyCrossInterpreterData.data type.

**Comment by pablogsal** [comment]
Buildbots have detected a leak of memory blocks in this PR. I have oppened:

https://bugs.python.org/issue35984

I do not have time to look into this until this night.

**Pull request merged**